src/Data/Bit/Immutable.hs

```diff
@@ -311,44 +312,94 @@ zipBits
   -> U.Vector Bit
   -> U.Vector Bit
   -> U.Vector Bit
-zipBits _ (BitVec _ 0 _) _ = U.empty
-zipBits _ _ (BitVec _ 0 _) = U.empty
+zipBits f = \xs ys -> case (xs, ys) of
+  (BitVec _ 0 _, !_) -> U.empty
+  (_, BitVec _ 0 _) -> U.empty
 #if UseSIMD
-zipBits f (BitVec 0 l1 arg1) (BitVec 0 l2 arg2) = runST $ do
-    let l = l1 `min` l2
+  (BitVec 0 l1 arg1, BitVec 0 l2 arg2) -> runST $ do
+    let
+        l = noinlineMin l1 l2
         w = nWords l
         b = wordsToBytes w
     brr <- newByteArray b
-    let ff = unBit $ f (Bit False) (Bit False)
-        ft = unBit $ f (Bit False) (Bit True)
-        tf = unBit $ f (Bit True)  (Bit False)
-        tt = unBit $ f (Bit True)  (Bit True)
-    case (ff, ft, tf, tt) of
-      (False, False, False, False) -> setByteArray brr 0 w (zeroBits :: Word)
-      (False, False, False, True)  -> ompAnd  brr arg1 arg2 b
-      (False, False, True,  False) -> ompAndn brr arg1 arg2 b
-      (False, False, True,  True)  -> copyByteArray brr 0 arg1 0 b
-      (False, True,  False, False) -> ompAndn brr arg2 arg1 b
-      (False, True,  False, True)  -> copyByteArray brr 0 arg2 0 b
-      (False, True,  True,  False) -> ompXor  brr arg1 arg2 b
-      (False, True,  True,  True)  -> ompIor  brr arg1 arg2 b
-      (True,  False, False, False) -> ompNior brr arg1 arg2 b
-      (True,  False, False, True)  -> ompXnor brr arg1 arg2 b
-      (True,  False, True,  False) -> ompCom  brr arg2      b
-      (True,  False, True,  True)  -> ompIorn brr arg1 arg2 b
-      (True,  True,  False, False) -> ompCom  brr arg1      b
-      (True,  True,  False, True)  -> ompIorn brr arg2 arg1 b
-      (True,  True,  True,  False) -> ompNand brr arg1 arg2 b
-      (True,  True,  True,  True)  -> setByteArray brr 0 w (complement zeroBits :: Word)
+    case unBitsy $ f (Bitsy 0b0011) (Bitsy 0b0101) of
+     0b0000 -> setByteArray brr 0 w (zeroBits :: Word)
+     0b0001 -> ompAnd  brr arg1 arg2 b
+     0b0010 -> ompAndn brr arg1 arg2 b
+     0b0011 -> copyByteArray brr 0 arg1 0 b
+     0b0100 -> ompAndn brr arg2 arg1 b
+     0b0101 -> copyByteArray brr 0 arg2 0 b
+     0b0110 -> ompXor  brr arg1 arg2 b
+     0b0111 -> ompIor  brr arg1 arg2 b
+     0b1000 -> ompNior brr arg1 arg2 b
+     0b1001 -> ompXnor brr arg1 arg2 b
+     0b1010 -> ompCom  brr arg2      b
+     0b1011 -> ompIorn brr arg1 arg2 b
+     0b1100 -> ompCom  brr arg1      b
+     0b1101 -> ompIorn brr arg2 arg1 b
+     0b1110 -> ompNand brr arg1 arg2 b
+     _0b1111 -> setByteArray brr 0 w (complement zeroBits :: Word)
     BitVec 0 l <$> unsafeFreezeByteArray brr
 #endif
-zipBits f xs ys = runST $ do
-  let n = min (U.length xs) (U.length ys)
-  zs <- MU.new n
-  forM_ [0, wordSize .. n - 1] $ \i ->
-    writeWord zs i (f (indexWord xs i) (indexWord ys i))
-  U.unsafeFreeze zs
-{-# INLINABLE zipBits #-}
+  _ -> runST $ do
+    let n = noinlineMin (U.length xs) (U.length ys)
```

Comment:
Could it be literally [`noinline`](https://hackage.haskell.org/package/base-4.18.0.0/docs/GHC-Exts.html#v:noinline) `min`?

Reply:
It looks like that makes trouble for unboxing.

Reply:
What I really want is

```haskell
lateInlineMin :: Int -> Int -> Int
lateInlineMin (I# x) (I# y) = I# (lateInline# $ case min (I# x) (I# y) of I# m -> m)
```

where `lateInline#` is a hypothetical magic function that delays inlining to Core Prep without either messing with demand or requiring a lifted argument (as `lazy` does).